### PR TITLE
Migrating flux checkpointing to hf api

### PR DESCRIPTION
**Pull request by wesleytruong**
Migrating the checkpointing of the flux dataset to use huggingface's faster api for IterableDatasets.

### CODE REVIEW

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 1]
```diff
@@ -126,6 +126,13 @@ class TextToImageDatasetConfig:
         ),
         data_processor=_cc12m_wds_data_processor,
     ),
+    "cc12m-test-iterable": TextToImageDatasetConfig(
+        path="torchtitan/experiments/flux/tests/assets/cc12m_test",
+        loader=lambda path: load_dataset(
+            path, split="train", data_files={"train": "*tar"}
+        ).to_iterable_dataset(num_shards=4),
+        data_processor=_cc12m_wds_data_processor,
+    ),
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]: let's put this into the test file instead of here

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 2]
```diff
@@ -4,96 +4,94 @@
...
+                    [
+                        f"--experimental.custom_args_module={path}",
+                        "--training.img_size",
+                        str(256),
+                        "--training.dataset",
+                        dataset_name,
+                        "--training.local_batch_size",
+                        str(batch_size),
+                        "--training.seed",
+                        "0",
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: I don't think we need this, as you specify seed for every step below

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 3]
```diff
@@ -4,96 +4,94 @@
...
-        for dataset_name in ["cc12m-test"]:
-            self._test_flux_dataloader(dataset_name)
+        # The test checks for the correct tensor shapes during the first num_steps
+        # The next num_steps ensure the loaded from checkpoint dataloader generates tokens and labels correctly
+        for world_size in [2]:
+            for rank in range(world_size):
+                dataset_name = "cc12m-test-iterable"
+                batch_size = 1
+
+                num_steps = 15
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:43]: Let's add a TODO here stating the limitation -- something like "if num_steps * batch_size is larger than the number of samples in the dataset, then the test would fail, due to issue ..."

**Review by tianyu-l — COMMENTED** [review]
Looks good, please address remaining comments. Also please rebase to latest main as there seem be conflicts.

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 4]
```diff
@@ -194,13 +194,13 @@ def __init__(
         self._all_samples: list[dict[str, Any]] = []
 
     def _get_data_iter(self):
-        if isinstance(self._data, Dataset) and self._sample_idx == len(self._data):
-            return iter([])
+        if isinstance(self._data, Dataset):
```
Comment (wwwjn) [inline torchtitan/experiments/flux/dataset/flux_dataset.py:197]: What if the self._data is not an instance of Dataset, will iter(self._data) returns data from beginning?

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 5]
```diff
@@ -32,16 +57,10 @@ def test_load_dataset(self):
                         dataset_name,
                         "--training.local_batch_size",
                         str(batch_size),
-                        "--training.seed",
-                        "0",
-                        "--training.classifer_free_guidance_prob",
-                        "0.447",
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: why removing this? I think this is indeed used in the code.

Also there is a misspell "classifier", could you help fix it globally?

**Review by tianyu-l — COMMENTED** [review]
One more nit comment

File: torchtitan/experiments/flux/dataset/flux_dataset.py (hunk 4 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/dataset/flux_dataset.py:197]: no, because `load_state_dict` below will take care of it. I agree this code structure difference between iterable and map-style is not perfect for read. Let's see if we can improve them later.
